cpp/src/arrow/io/memory.h

```diff
@@ -72,6 +72,13 @@ class ARROW_EXPORT BufferOutputStream : public OutputStream {
 
   int64_t capacity() const { return capacity_; }
 
+  /// \brief Return reference to current in-progress buffer. May
+  /// become invalid after any Write calls
+  const Buffer& buffer() const { return *buffer_; }
```

Comment:
This doesn't seem right at all. Why does Parquet need this?

Reply:
Parquet reuses memory on the encoding hot path. Otherwise we have to call `Finish` and then re-initialize the BufferOutputStream

Reply:
I can also rewrite this code to use BufferBuilder or something lower-level

Reply:
The `BufferOutputStream` could have decided to flush the buffer or part of it... I don't think you can rely on this to always give you all the data you wrote before.

Reply:
I went ahead and rewrote the code in question to use `arrow::BufferBuilder`. I added a `BufferBuilder::Rewind` method